registry.go

```diff
@@ -0,0 +1,328 @@
+// limitations under the License.
+
+// TODO: split into separate package
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+const (
+	// Capacity for the channel to collect metrics and descriptors.
+	capMetricChan = 1000
+	capDescChan   = 10
+)
+
+var (
+	registry = NewMetricsRegistry()
+)
+
+// For statsd all our metrics will be both a Metric and a collector -- to make life easier
+// we'll define an interface
+type CollectorMetric interface {
+	prometheus.Metric
+	prometheus.Collector
+}
+
+// TODO: remove once https://github.com/prometheus/client_golang/pull/326 is merged
+type Metric struct {
+	// TODO: remove once these have getters upstream
+	Name string
+	Help string
+
+	M CollectorMetric
+}
+
+// TODO: change back to using the regular mechanisms to determine collisions after https://github.com/prometheus/client_golang/pull/326 is merged
+// then we can actually implement the registerer interface
+type MetricsRegistry struct {
+	mtx           sync.RWMutex
+	metricsByID   map[uint64]*Metric // ID is a hash of the descIDs.
+	metricsByDesc map[*prometheus.Desc]*Metric
+}
+
+// NewRegistry creates a new vanilla Registry without any Collectors
+// pre-registered.
+func NewMetricsRegistry() *MetricsRegistry {
+	return &MetricsRegistry{
+		metricsByID:   map[uint64]*Metric{},
+		metricsByDesc: map[*prometheus.Desc]*Metric{},
+	}
+}
+
+// TODO: (deosn't work now) Register implements Registerer.
+func (r *MetricsRegistry) Register(metricID uint64, m *Metric) error {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if _, exists := r.metricsByID[metricID]; exists {
+		/* TODO?
+		return prometheus.AlreadyRegisteredError{
+			ExistingCollector: existing,
+			NewCollector:      c,
+		}
+		*/
+		return fmt.Errorf("Duplicate metric hash")
+	} else {
+		r.metricsByID[metricID] = m
+		r.metricsByDesc[m.M.Desc()] = m
+		return nil
+	}
+}
+
+// TODO: (deosn't work now) Unregister implements Registerer.
+func (r *MetricsRegistry) Unregister(metricID uint64) bool {
+	r.mtx.RLock()
+	defer r.mtx.RUnlock()
+	if existingMetric, exists := r.metricsByID[metricID]; exists {
+		delete(r.metricsByID, metricID)
+		delete(r.metricsByDesc, existingMetric.M.Desc())
+		return true
+	}
+	return false
+}
+
+// Gather implements Gatherer.
+func (r *MetricsRegistry) Gather() ([]*dto.MetricFamily, error) {
+	var (
+		metricChan = make(chan prometheus.Metric, capMetricChan)
+		wg         sync.WaitGroup
+		errs       MultiError // The collected errors to return in the end.
+	)
+
+	r.mtx.RLock()
+	metricFamiliesByName := make(map[string]*dto.MetricFamily)
+
+	// Scatter.
+	// (Collectors could be complex and slow, so we call them all at once.)
+	wg.Add(len(r.metricsByID))
+	go func() {
+		wg.Wait()
+		close(metricChan)
+	}()
+	for _, metric := range r.metricsByID {
+		go func(collector prometheus.Collector) {
```

Comment:
As pointed out in https://github.com/prometheus/statsd_exporter/issues/76#issuecomment-359783877 this causes a huge spike in memory consumption because it spawns a goroutine for every single time series.

prometheus/client_golang#370 is a potential fix for that.